dora/core/common/src/main/java/alluxio/uri/UfsUrl.java

```diff
@@ -0,0 +1,414 @@
+    }
+    sb.append(mProto.getAuthority());
+    if (!mProto.getAuthority().isEmpty()) {
+      sb.append(UfsUrl.PATH_SEPARATOR);
+    }
+    List<String> pathComponents = mProto.getPathComponentsList();
+    for (int i = 0; i < pathComponents.size(); i++) {
+      if (pathComponents.get(i).isEmpty())  {
+        continue;
+      }
+      sb.append(pathComponents.get(i));
+      if (i != pathComponents.size() - 1) {
+        sb.append(UfsUrl.PATH_SEPARATOR);
+      }
+    }
+    return sb.toString();
+  }
+
+  /**
+   * @return hashCode of {@link UfsUrl}
+   */
+  // TODO(Tony Sun): determine how to modify it.
+  public int hashCode() {
+    String s = asString();
+    int len = s.length();
+    int hash = 0;
+    for (int i = 0; i < len; i++) {
+      hash = (hash + s.charAt(i)) % Integer.MAX_VALUE;
+    }
+    return hash;
+  }
+
+  /**
+   * determine whether object o is equal to current UfsUrl.
+   *
+   * @param o an object
+   * @return true if equal, false if not equal
+   */
+  public boolean equals(Object o) {
+    if (this == o)  {
+      return true;
+    }
+    if (!(o instanceof UfsUrl)) {
+      return false;
+    }
+    UfsUrl that = (UfsUrl) o;
+    return mProto.equals(that.mProto);
+  }
+
+  /**
+   * Get parent UfsUrl of current UfsUrl.
+   *
+   * @return parent UfsUrl
+   */
+  // TODO(Jiacheng Liu): try to avoid the copy by a RelativeUrl class
+  public UfsUrl getParentURL() {
+    List<String> pathComponents = mProto.getPathComponentsList();
+    return new UfsUrl(UfsUrlMessage.newBuilder()
+        .setScheme(mProto.getScheme())
+        .setAuthority(mProto.getAuthority())
+        // TODO(Jiacheng Liu): how many copies are there. Improve the performance in the future.
+        .addAllPathComponents(pathComponents.subList(0, pathComponents.size() - 1)).build());
+  }
+
+  /**
+   * Get a child UfsUrl of current UfsUrl.
+   *
+   * @param childName a child file/directory of this object
+   * @return a child UfsUrl
+   */
+  // TODO(Jiacheng Liu): try to avoid the copy by a RelativeUrl class
+  public UfsUrl getChildURL(String childName) {
+    List<String> pathComponents = mProto.getPathComponentsList();
+    return new UfsUrl(UfsUrlMessage.newBuilder()
+        .setScheme(mProto.getScheme())
+        .setAuthority(mProto.getAuthority())
+        .addAllPathComponents(pathComponents).addPathComponents(childName).build());
+  }
+
+  /**
+   * Return the full path by connect the pathComponents list.
+   *
+   * @return a full path string
+   */
+  public String getFullPath() {
+    String partPath = Strings.join(mProto.getPathComponentsList(), PATH_SEPARATOR.charAt(0));
+    return partPath.startsWith("/") ? partPath : "/" + partPath;
+  }
+
+  /**
+   * Translate current UfsUrl object to an AlluxioURI object.
+   *
+   * @return a corresponding AlluxioURI object
+   */
+  public AlluxioURI toAlluxioURI() {
+    return new AlluxioURI(mProto.getScheme(),
+        Authority.fromString(mProto.getAuthority()), getFullPath());
```

Comment:
why do you need `Authority.fromString()`?

Reply:
AlluxioURI is not support ctor like AlluxioURI(String, String, String), but it support AlluxioURI(String, Authority, String).